src/libs/maintaining-std.md

```diff
@@ -187,7 +187,40 @@ Changes to collection internals may affect the order their items are dropped in.
 
 ### Is there a manual `Drop` implementation?
 
-Generic types that manually implement `Drop` should consider whether a `#[may_dangle]` attribute is appropriate. The [Nomicon][dropck] has some details on what `#[may_dangle]` is all about.
+A generic `Type<T>` that manually implements `Drop` should consider whether a `#[may_dangle]` attribute is appropriate on `T`. The [Nomicon][dropck] has some details on what `#[may_dangle]` is all about.
+
+If a generic `Type<T>` has a manual drop implementation that may also involve dropping `T` then dropck needs to know about it. If `Type<T>`'s ownership of `T` is expressed through types that don't drop `T` themselves such as `ManuallyDrop<T>`, `*mut T`, or `MaybeUninit<T>` then `Type<T>` also [needs a `PhantomData<T>` field][RFC 0769 PhantomData] to tell dropck that `T` may be dropped. Types in the standard library that use the internal `Unique<T>` pointer type don't need a `PhantomData<T>` marker field. That's taken care of for them by `Unique<T>`.
+
+As a real-world example of where this can go wrong, consider an `OptionCell<T>` that looks something like this:
+
+```rust
+struct OptionCell<T> {
+    is_init: bool,
+    value: MaybeUninit<T>,
+}
+
+impl<T> Drop for OptionCell<T> {
+    fn drop(&mut self) {
+        if self.is_init {
+            // Safety: The cell is being dropped, so it can't be accessed again.
+            let _ = unsafe { self.value.read() };
```

Comment:
```suggestion
            drop(unsafe { self.value.read() });
```

Reply:
I think I like the version without `drop` explicitly written, to show how subtle the bug can be. (Just like in the real problem with `SyncOnceCell`.) (But don't feel strongly about it.)

Reply:
I don't think we should write hard-to-read code in the docs just because there is hard-to-read code in the wild.^^ IMO if the real code dropped via `let _ =`, it should be changed, too.

However, I take your point insofar as it is worth pointing out in the text that a call to drop can be implicit.